lib/coordinator.go

```diff
@@ -0,0 +1,285 @@
+	return c.queue.Chan()
+}
+
+// Completed sends one or more constructed resources to the coordinator
+func (c *Coordinator) Completed(rsc ...*Resource) error {
+
+	// handle any global state changes that may result from completed work
+	// TODO - finish
+	go func() {
+		// for _, resc := range c.cfg.BackoffResponseCodes {
+		// 	if res.StatusCode == resc {
+		// 		log.Infof("encountered %d response. backing off", resc)
+		// 		c.setCrawlDelay(c.crawlDelay + ((time.Duration(c.cfg.CrawlDelayMilliseconds) * time.Millisecond) / 2))
+		// 	}
+		// }
+		// if c.finished == c.cfg.StopAfterEntries {
+		// 	stop <- true
+		// }
+	}()
+
+	// handle resources and create a deduplicated map
+	// of unique candidate urls from all responses
+	links := map[string]bool{}
+	for _, r := range rsc {
+		if err := c.dequeue(r); err != nil {
+			log.Debugf("error dequing url: %s: %s", r.URL, err.Error())
+		}
+
+		for _, l := range r.Links {
+			if c.urlStringIsCandidate(l) {
+				links[l] = true
+			}
+		}
+	}
+
+	for url := range links {
+		r, err := c.frs.Get(url)
+		if err != nil {
+			log.Debugf("err getting url: %s: %s", url, err.Error())
+		}
+		if r == nil {
+			c.enqueue(&Request{URL: url})
+		}
+	}
+
+	return nil
+}
+
+func (c *Coordinator) enqueue(rs ...*Request) {
+	for _, r := range rs {
+		if c.stopping {
+			r.Status = RequestStatusFailed
+			c.frs.Put(r)
+			continue
+		}
+
+		log.Debugf("enqueue: %s", r.URL)
+		r.Status = RequestStatusQueued
+		c.frs.Put(r)
+		c.queue.Push(r)
+	}
+}
+
+func (c *Coordinator) dequeue(rsc *Resource) error {
+	fr, err := c.frs.Get(rsc.URL)
+	if err == ErrNotFound {
+		fr = &Request{URL: rsc.URL}
+	} else if err != nil {
+		log.Debugf("err getting url: %s: %s", rsc.URL, err.Error())
+		return err
+	}
+
+	fr.PrevResStatus = rsc.Status
+	fr.AttemptsMade++
+
+	if c.okResponseStatus(fr.PrevResStatus) {
+		log.Debugf("dequeue: %s", fr.URL)
+		c.finished++
+		c.urlsWritten++
+		fr.Status = RequestStatusDone
+		// send completed records to each handler
+		for _, h := range c.handlers {
+			go h.HandleResource(rsc)
+		}
+		return nil
+	}
+
+	if fr.AttemptsMade <= c.cfg.MaxAttempts {
+		c.enqueue(fr)
+		return nil
+	}
+
+	fr.Status = RequestStatusFailed
+	return c.frs.Put(fr)
+}
+
+func (c *Coordinator) setCrawlDelay(d time.Duration) {
+	c.crawlDelay = d
+	for _, c := range c.workers {
+		c.SetDelay(d)
+	}
+	log.Infof("crawler delay is now: %f seconds", c.crawlDelay.Seconds())
+	log.Infof("crawler request frequency: ~%f req/minute", float64(len(c.workers))/c.crawlDelay.Minutes())
+}
+
+// urlStringIsCandidate scans the slice of crawlingURLS to see if we should GET
+// the passed-in url
+func (c *Coordinator) urlStringIsCandidate(rawurl string) bool {
+	for _, ignore := range c.cfg.IgnorePatterns {
+		if strings.Contains(rawurl, ignore) {
+			return false
+		}
+	}
+
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return false
+	}
+	for _, d := range c.domains {
+		if u.Path != "" && !strings.Contains(u.Path, d.Path) {
```

Comment:
Two notes here:

1. We should be checking whether `d.Path` is present rather than whether `u.Path` is present, otherwise `http://example.com` is a valid candidate even if our domain list has `http://example.com/sub/path`.

2. Should probably be `strings.HasPrefix()` instead of `strings.Contains()`.

Reply:
1. That's what this is doing 😄
2. Agreed. Will fix

Reply:
> 1. That's what this is doing 😄

Ummmm, am I misunderstanding? I’m thinking this situation:

- Configured domains: `["http://example.com/sub/path"]`
- Candidate URL: `http://example.com`

This function is supposed to reject that candidate, right? But it won’t because `u.Path == ""`. It’ll short-circuit before the prefix check.